### DISCUSSION TITLE
Implementation of cross-end and cross-window file system service based on OpenSumi

### DISCUSSION THREAD

**Discussion by erha19** [opensumi/core#1086] [General]
The meaning of this topic can be simply understood as "How to implement cross-window file operations through OpenSumi".

On the OpenSumi, now we can use `Cmd+C / Cmd+V` or drag some files from one position to the others on the same IDE window, but we can't use this function while copy or drag some files from one window to the other window.

You can see the dragging logic here: [file-tree-dnd.service.ts#L68](https://github.com/opensumi/core/blob/3340fa89eab63413a00335f52a81d6cfa20dd2e5/packages/file-tree-next/src/browser/services/file-tree-dnd.service.ts#L68), it will bind into all TreeNodes on the fileTree view, and use the mv API defined in the FileService to move the files, see [file-tree-api.service.ts#L166](https://github.com/opensumi/core/blob/3340fa89eab63413a00335f52a81d6cfa20dd2e5/packages/file-tree-next/src/browser/services/file-tree-api.service.ts#L166), you can also see the similar codes in the case of Copy-Paste.

There are some ideas:

On the VS Code, it use an `clipboardService` to save the resources list, and implement difference service with the Web and Electorn, you can see some code here:

- Copy files [explorerService.ts#L227](https://github.com/microsoft/vscode/blob/7fc8af4e2d8f50a3950b809b9beab86fa5135f29/src/vs/workbench/contrib/files/browser/explorerService.ts#L227) 
- Electron implement [clipboardService.ts](https://github.com/microsoft/vscode/blob/01472963d7b9f963feb5932d7a9276f4dc085619/src/vs/workbench/services/clipboard/electron-sandbox/clipboardService.ts)
- Web implement [clipboardService.ts](https://github.com/microsoft/vscode/blob/6b9dde1cd3686e1b95b90388a0995187c1c63020/src/vs/platform/clipboard/browser/clipboardService.ts)

On the Web, it can just copy-paste files on the same window, 
This is good enough, but it can be done better, e.g. you can also use some shared storage on the web to handle this, such as localstorage.

On the Electron, it use `this.nativeHostService.writeClipboardBuffer` function to share and update the copy resources list , so you can copy-paste files cross-window.

Also, our FileService is not running very well, there will be some redundant or wrong events, we also hope to consider and deal with these problems together.

**Comment by Rispectech** [comment]
Hey, I am rishabh. I talked to you about this issue earlier. so I have been browsing your codebase and researching this issue since Friday. 
So, I have tried to study clipboard API and seen the implementation of vs code. so some of the viable solutions are either we use clipboard API entirely to copy and paste the file or we can use clipboard API to copy the content in local storage and then write the file from there. 

can you guide me on how to proceed further and what can of example are you expecting before the proposal. Morever, there are 28 modules in the project. I have read the contributing guide but i am finding the code a bit overwhelming so can you guide me a bit from where to start?

**Reply by erha19** [reply]
@Rispectech I'm sorry to tell you the quota is full cause the project can pick only one student, we will consider prioritizing students who contribute to our framework, which we have asked @songhn233  to do, and he is also interested in doing it. Also, Alibaba will launch a new open-source summer camp too, you can see some topics here [ASOC2022](https://github.com/opensumi/core/issues/1103).

If you want to do that, you can firstly choose one or more issues with the `good first issue` label, and try to resolve it, this is an important factor in our selection of students.

**Comment by Rispectech** [comment]
@erha19  okay no problem 
i have a question instead is Opensumi 2nd project for ospp that was ***Using OpenSumi to implementing a Github CodeReview IDE*** taking students right now ?

**Reply by erha19** [reply]
As I know, the 2nd project `Using OpenSumi to implementing a Github CodeReview IDE` has been taking students too.

**Reply by Rispectech** [reply]
Oh okay thanks a lot  @erha19 . There repository was  not on GitHub . So can suggest me some ways to communicate with them ?

**Reply by erha19** [reply]
@Rispectech maybe you can mail your questions to the mentor of the project or join the DingTalk Group with the number 42100604.

**Comment by erha19** [comment]
Will be close by #1594.
